language-plutus-core/plutus-ir/Language/PlutusIR/Compiler/Error.hs

```diff
@@ -5,28 +5,74 @@
 {-# LANGUAGE TemplateHaskell        #-}
 {-# LANGUAGE TypeOperators          #-}
 {-# LANGUAGE UndecidableInstances   #-}
-module Language.PlutusIR.Compiler.Error (Error (..), AsError (..)) where
+module Language.PlutusIR.Compiler.Error
+    ( Error (..)
+    , TypeError (..)
+    , AsTypeError (..)
+    , AsError (..)
+    , PLC.Normalized (..)
+    , PLC.UnknownDynamicBuiltinNameError (..)
+    , throwingEither
+    ) where
 
 import qualified Language.PlutusCore        as PLC
 import qualified Language.PlutusCore.Pretty as PLC
+--import qualified Language.PlutusCore.Error as PLC (throwingEither)
+import qualified Language.PlutusIR as PIR
 
-import           Control.Exception
+import           Control.Exception hiding (TypeError)
 import           Control.Lens
 
 import qualified Data.Text                  as T
-import           Data.Text.Prettyprint.Doc  ((<+>))
-import qualified Data.Text.Prettyprint.Doc  as PP
+import Data.Text.Prettyprint.Doc  as PP
 import           Data.Typeable
+import GHC.Generics (Generic)
+import Control.Monad.Error.Lens
+import Control.Monad.Except
+
+
+-- | Lifts an 'Either' into an error context where we can embed the 'Left' value into the error.
+-- TODO: literally copy-pasted from PlutusCore.Error to avoid modifying the exposed modules/functions of language-plutus-core package
+throwingEither :: MonadError e m => AReview e t -> Either t a -> m a
+throwingEither r e = case e of
+    Left t  -> throwing r t
+    Right v -> pure v
+
+
+data TypeError uni ann
```

Comment:
Oh, it does actually have extra error types.

Reply:
Or we do the prismatic error handling thing.

Reply:
I think making `TypeError` extensible is simpler. For prismatic error handling we'll still need to parameterize `TypeError` by `term`, so we only save one type variable at the cost of introducing creepy looking constraints. Unless I'm missing something.

Reply:
I pushed an "extensible typeerrors" implementation; the only minor annoyance is that I had to introduce an extraneous `TypeMismatchPIR` data-constructor, because the PLC one refers to PLC.terms:

```
data TypeErrorExt uni ann =
    -- we have to duplicate this error because of PIR.Term
    TypeMismatchPIR ann
        (PIR.Term PLC.TyName PLC.Name uni ())
        (PLC.Type PLC.TyName uni ())
        (PLC.Normalized (PLC.Type PLC.TyName uni ()))
```

The good thing is that it cuts down a lot of the code in `pir/typecheck/internal.hs` file.

Reply:
You could parameterize the extensible `TypeError` by `term`. But the way it's currently written is perfectly fine as well.